src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,167 @@
+
+            var lastParameterIsCancellationToken = methodParameters.Length > 0 && methodParameters[^1].ParameterType == typeof(CancellationToken);
+
+            if (lastParameterIsCancellationToken)
+                Contract.ThrowIfFalse(request.Parameters.Length == methodParameters.Length - 1, $"The arguments list should contain every parameter for {request.Method} except the final CancellationToken.");
+            else
+                Contract.ThrowIfFalse(request.Parameters.Length == methodParameters.Length, $"The arguments list should contain every parameter for {request.Method}.");
+
+            var arguments = new object?[methodParameters.Length];
+
+            for (var i = 0; i < methodParameters.Length; i++)
+            {
+                // If the method we're calling accepts a cancellation token, we wouldn't have passed that, so add in a CancellationToken.None here. Although we could
+                // remove the cancellation from the underlying method, this keeps that support around should we need it.
+                if (i == methodParameters.Length - 1 && lastParameterIsCancellationToken)
+                    arguments[i] = CancellationToken.None;
+                else
+                    arguments[i] = request.Parameters[i].ToObject(methodParameters[i].ParameterType);
+            }
+
+            var result = method.Invoke(rpcTarget, arguments);
+
+            if (result is Task task)
+            {
+                await task.ConfigureAwait(false);
+
+                // If it's actually a Task<T> then get the result; we're looking at the declared return type because in some cases a method might
+                // return a Task<T> under the covers as a workaround for the lack of TaskCompletionSource on .NET Framework but we don't want to see
+                // that workaround since the result isn't intended to be seen.
+                if (method.ReturnType.IsConstructedGenericType)
+                {
+                    result = task.GetType().GetProperty("Result")!.GetValue(task);
+                }
+                else
+                {
+                    // It's just a simple Task so no result to return
+                    result = null;
+                }
+            }
+
+            response = new Response { Id = request.Id, Value = result is not null ? JToken.FromObject(result) : null };
+        }
+        catch (Exception e)
+        {
+            if (e is TargetInvocationException)
+                e = e.InnerException ?? e;
+
+            response = new Response { Id = request.Id, Exception = $"An exception of type {e.GetType()} was thrown: {e.Message}" };
+        }
+
+        var responseJson = JsonConvert.SerializeObject(response, JsonSettings.SingleLineSerializerSettings);
+
+#if DEBUG
+        // Assert we didn't put a newline in this, since if we did the receiving side won't know how to parse it
+        Contract.ThrowIfTrue(responseJson.Contains("\r") || responseJson.Contains("\n"));
```

Comment:
If the serialization above throws, or this throws - what happens to the exception? will we be able to know it happened?

Reply:
Practically it's hard to do much at this point: we did the serialization of the actual result data under the try/catch intentionally to report any failures there, so any failure at this point means we can't serialize anything or the input/output itself is dead. So not that we _couldn't_ try to reply with an error of some kind...how can I?

Reply:
Wondering if we should somehow log to a file or something.  Because right now if this fails will we ever know and figure out that this caused it?

Reply:
Any thrown exceptions here can potentially end up in the console output from the build host process, so we'll still see it (just admittedly in a harder to find place.) The challenge is we don't have a good other place to communicate problems like this.